content/Refactoring - Sebastian Herrera.md

```diff
@@ -0,0 +1,144 @@
+
+So, enough introduction, this is about what I learned from that book.
+
+## 1) WHY TO DO REFACTORING?
+Maybe the most important question to ask yourself before doing anything, is why should you do it, meaning what value does it give in return.
+
+These are a some of the main reasons why doing refactoring is recommended:
+- Refactoring improves the design of software.
+  - Creating software that has a good architecture isn't something that can be done first time. The design we believe that best fits today, might not be the one that best fits next week.
+  - Without refactoring the design of the program will decay over time.
+- Refactoring makes software easier to understand.
+  - *“Any fool can write code that a computer can understand. Good programmers write code that humans can understand.”* - Martin Fowler
+- Refactoring makes you find bugs.
+  - When the code is more simple, it is easier to understand, easier to debug, and also easier to detect non-considered cases.
+- Refactoring helps you program faster.
+  - Without a good design, you can progress quickly for a while, but soon the poor design starts to slow you down.
+
+**Conclusion: Refactoring lowers the cost of enhancements.**
+
+## 2) WHEN TO DO REFACTORING?
+Refactoring isn't a special task that will show up in a project plan, so when should you do it?
+- When you need to add a new feature to a codebase, look at the existing code and consider whether it's structured in such a way to make the new change straightforward.
+  - If it isn't, then refactor the existing code to make this new addition easy.
+  - Once that change is done, then add the new feature.
+- After you add a feature and get it working, you can a review the resulting code, that while it works, it may not be as clear as it could be.
+  - If it isn't , then refactor it into a better shape so that when you (or someone else) return to this code in a few weeks time, you won't have to spend time puzzling out how this code works.
+- If you struggle to understand the code, refactor it so you won't have to struggle again next time you look at it.
+
+**Conclusion: Refactoring is a part of day-to-day programming.**
+
+## 3) CODE SMELLS
+Probably you’ve heard about them, and for sure you have already seen tons of them. Let’s see their characteristics.
+- A code smell is a surface indication that usually corresponds to a deeper problem in the system.
+  - Firstly a smell is by definition something that's quick to spot - or sniffable.
+  - A long method is a good example of this - just looking at the code and my nose twitches if I see more than a dozen lines.
+- The second is that smells don't always indicate a problem.
+  - Some long methods are just fine. You have to look deeper to see if there is an underlying problem there - smells aren't inherently bad on their own - they are often an indicator of a problem rather than the problem themselves.
```

Comment:
```suggestion
  - Some long methods are just fine. You have to look deeper to see if there is an underlying problem - smells aren't inherently bad on their own - they are often an indicator of a problem rather than the problem themselves.
```

BTW: not sure which one is better, feel free to ignore this 😄